// @adobe/cq-angular-editable-components/esm2015/lib/layout/aem-model-provider/aem-model-provider.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/*
 * ADOBE CONFIDENTIAL
 *
 * Copyright 2018 Adobe Systems Incorporated
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Adobe Systems Incorporated and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Adobe Systems Incorporated and its
 * suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe Systems Incorporated.
 */
import { Component, Input, NgZone } from '@angular/core';
import { ModelManager } from "@adobe/cq-spa-page-model-manager";
/**
 * The current component is responsible for providing access to the ModelManager and the model of a component
 */
export class AEMModelProviderComponent {
    /**
     * @param {?} ngZone
     */
    constructor(ngZone) {
        this.ngZone = ngZone;
    }
    /**
     * Updates the item data
     * @return {?}
     */
    updateItem() {
        ModelManager.getData({ path: this.cqPath }).then(model => {
            this.ngZone.run(() => {
                this.cqItem = model;
            });
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        ModelManager.addListener(this.cqPath, this.updateItem.bind(this));
    }
    /**
     * @return {?}
     */
    ngDestroy() {
        ModelManager.removeListener(this.cqPath, this.updateItem.bind(this));
    }
}
AEMModelProviderComponent.decorators = [
    { type: Component, args: [{
                selector: '[aemModelProvider]',
                template: `<ng-container [aemComponent] [cqItem]='cqItem' [cqPath]='cqPath' [itemName]='itemName'></ng-container>`
            },] },
];
/** @nocollapse */
AEMModelProviderComponent.ctorParameters = () => [
    { type: NgZone, },
];
AEMModelProviderComponent.propDecorators = {
    "cqPath": [{ type: Input },],
    "cqItem": [{ type: Input },],
    "itemName": [{ type: Input },],
    "aemModelProvider": [{ type: Input },],
};
function AEMModelProviderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AEMModelProviderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AEMModelProviderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    AEMModelProviderComponent.propDecorators;
    /**
     * Path to the model associated with the current instance of the component
     * @type {?}
     */
    AEMModelProviderComponent.prototype.cqPath;
    /**
     * Model item associated with the current model provider component
     * @type {?}
     */
    AEMModelProviderComponent.prototype.cqItem;
    /**
     * Name of the item associated with the current model provider component
     * @type {?}
     */
    AEMModelProviderComponent.prototype.itemName;
    /** @type {?} */
    AEMModelProviderComponent.prototype.aemModelProvider;
    /** @type {?} */
    AEMModelProviderComponent.prototype.ngZone;
}
