// @adobe/cq-angular-editable-components/esm2015/lib/layout/aem-responsivegrid/aem-responsivegrid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/*
 * ADOBE CONFIDENTIAL
 *
 * Copyright 2018 Adobe Systems Incorporated
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Adobe Systems Incorporated and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Adobe Systems Incorporated and its
 * suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe Systems Incorporated.
 */
import { Component, Input } from '@angular/core';
import { AEMContainerComponent } from '../aem-container/aem-container.component';
import { Constants } from '../constants';
const /** @type {?} */ PLACEHOLDER_CLASS_NAMES = ' aem-Grid-newComponent';
const /** @type {?} */ RESPONSIVE_GRID_TYPE = 'wcm/foundation/components/responsivegrid';
/**
 * The current class carries the base presentational logic of the AEM Layout Container (aka. Responsive grid)
 */
export class AEMResponsiveGridComponent extends AEMContainerComponent {
    /**
     * Returns the column class names for a given column
     * @param {?} itemKey - The key of the column item
     * @return {?}
     */
    getColumnClassNames(itemKey) {
        return this.columnClassNames && this.columnClassNames[itemKey];
    }
    /**
     * Returns the placeholder classes
     * @return {?}
     */
    getPlaceholderClassNames() {
        return super.getPlaceholderClassNames() + PLACEHOLDER_CLASS_NAMES;
    }
    /**
     * Returns the class names of the responsive grid based on the data from the cqModel
     * @return {?}
     */
    getHostClassNames() {
        let /** @type {?} */ classNames = super.getHostClassNames();
        if (this.classNames) {
            classNames += ' ' + (this.classNames || '');
        }
        return classNames + ' ' + this.gridClassNames;
    }
    /**
     * Returns the aggregated path of this container path and the provided path
     *
     * @param {?} path - the provided path to aggregate with the container path
     * @return {?}
     */
    getAttrDataPath(path) {
        let /** @type {?} */ item = this.getItem(path);
        if (item && item[Constants.TYPE_PROP] === RESPONSIVE_GRID_TYPE) {
            // We don't want to add the path for the wrapper for a reponsivegrid
            // The reponsivegrid adds the path on it's own
            return null;
        }
        return this.getDataPath(path);
    }
}
AEMResponsiveGridComponent.decorators = [
    { type: Component, args: [{
                selector: 'aem-responsivegrid',
                host: {
                    '[class]': 'hostClasses',
                    '[attr.data-cq-data-path]': 'cqPath'
                },
                template: `<ng-container *ngFor="let itemKey of cqItemsOrder">
  <div [aemModelProvider] [cqItem]='getItem(itemKey)' [cqPath]='getDataPath(itemKey)' [itemName]='itemKey' [class]="getColumnClassNames(itemKey)" [attr.data-cq-data-path]='getAttrDataPath(itemKey)'>
  </div>
</ng-container>
<div *ngIf="isInEditMode" [attr.data-cq-data-path]="placeholderPath" [class]="getPlaceholderClassNames()"></div>


`,
            },] },
];
/** @nocollapse */
AEMResponsiveGridComponent.propDecorators = {
    "gridClassNames": [{ type: Input },],
    "columnClassNames": [{ type: Input },],
    "classNames": [{ type: Input },],
    "columnCount": [{ type: Input },],
};
function AEMResponsiveGridComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AEMResponsiveGridComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AEMResponsiveGridComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    AEMResponsiveGridComponent.propDecorators;
    /**
     * Class names associated with the current responsive grid
     * @type {?}
     */
    AEMResponsiveGridComponent.prototype.gridClassNames;
    /**
     * Map of class names corresponding to each child of the current responsive grid
     * @type {?}
     */
    AEMResponsiveGridComponent.prototype.columnClassNames;
    /**
     * Class names of the current component
     * @type {?}
     */
    AEMResponsiveGridComponent.prototype.classNames;
    /**
     * Current number of columns of the grid
     * @type {?}
     */
    AEMResponsiveGridComponent.prototype.columnCount;
}
