// @adobe/cq-angular-editable-components/esm5/lib/routing/AemPageRouteReuseStrategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Implements RouteReuseStrategy to customize route reuse.
 */
var /**
 * Implements RouteReuseStrategy to customize route reuse.
 */
AemPageRouteReuseStrategy = /** @class */ (function () {
    function AemPageRouteReuseStrategy() {
    }
    /** Determines if this route (and its subtree) should be detached to be reused later. */
    /**
     * Determines if this route (and its subtree) should be detached to be reused later.
     * @param {?} route
     * @return {?}
     */
    AemPageRouteReuseStrategy.prototype.shouldDetach = /**
     * Determines if this route (and its subtree) should be detached to be reused later.
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return false;
    };
    /** Not storing deteached route. */
    /**
     * Not storing deteached route.
     * @param {?} route
     * @param {?} detachedTree
     * @return {?}
     */
    AemPageRouteReuseStrategy.prototype.store = /**
     * Not storing deteached route.
     * @param {?} route
     * @param {?} detachedTree
     * @return {?}
     */
    function (route, detachedTree) { };
    /** Determines if this route (and its subtree) should be reattached. */
    /**
     * Determines if this route (and its subtree) should be reattached.
     * @param {?} route
     * @return {?}
     */
    AemPageRouteReuseStrategy.prototype.shouldAttach = /**
     * Determines if this route (and its subtree) should be reattached.
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return false;
    };
    /** Retrieves the previously stored route. */
    /**
     * Retrieves the previously stored route.
     * @param {?} route
     * @return {?}
     */
    AemPageRouteReuseStrategy.prototype.retrieve = /**
     * Retrieves the previously stored route.
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return null;
    };
    /** Determines if a route should be reused */
    /**
     * Determines if a route should be reused
     * @param {?} future
     * @param {?} curr
     * @return {?}
     */
    AemPageRouteReuseStrategy.prototype.shouldReuseRoute = /**
     * Determines if a route should be reused
     * @param {?} future
     * @param {?} curr
     * @return {?}
     */
    function (future, curr) {
        return false;
    };
    return AemPageRouteReuseStrategy;
}());
/**
 * Implements RouteReuseStrategy to customize route reuse.
 */
export { AemPageRouteReuseStrategy };
