// @adobe/cq-angular-editable-components/esm2015/lib/layout/aem-container/aem-container.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/*
 * ADOBE CONFIDENTIAL
 *
 * Copyright 2018 Adobe Systems Incorporated
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Adobe Systems Incorporated and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Adobe Systems Incorporated and its
 * suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe Systems Incorporated.
 */
import { Component, Input } from '@angular/core';
import { Constants } from '../constants';
import { Utils } from "../utils";
const /** @type {?} */ PLACEHOLDER_CLASS_NAMES = Constants.NEW_SECTION_CLASS_NAMES;
const /** @type {?} */ PLACEHOLDER_ITEM_NAME = '*';
const /** @type {?} */ CONTAINER_CLASS_NAMES = 'aem-container';
/**
 * The current component provides the base presentational logic common to containers such as a grid or a page.
 * Container have in common the notion of item holders. Items are represented in the model by the fields _:items_ and _:itemsOrder_
 */
export class AEMContainerComponent {
    constructor() {
        /**
         * Path to the model associated with the current instance of the component
         */
        this.cqPath = '';
        /**
         * Key of the model structure
         */
        this.modelName = '';
    }
    /**
     * Returns weather of not we are in the editor
     * @return {?}
     */
    get isInEditMode() {
        return Utils.isInEditor();
    }
    /**
     * Returns the aggregated path of this container path and the provided path
     *
     * @param {?} path - the provided path to aggregate with the container path
     * @return {?}
     */
    getDataPath(path) {
        return this.cqPath ? this.cqPath + '/' + path : path;
    }
    /**
     * Returns the item data from the cqModel
     *
     * @param {?} itemKey - the itemKey to look for in the items.
     * @return {?}
     */
    getItem(itemKey) {
        return this.cqItems && this.cqItems[itemKey];
    }
    /**
     * Returns the class names of the container based on the data from the cqModel
     * @return {?}
     */
    getHostClassNames() {
        return CONTAINER_CLASS_NAMES;
    }
    /**
     * @return {?}
     */
    get hostClasses() {
        return this.getHostClassNames();
    }
    /**
     * Returns the placeholder classes
     * @return {?}
     */
    getPlaceholderClassNames() {
        return PLACEHOLDER_CLASS_NAMES;
    }
    /**
     * Returns the placeholder path
     * @return {?}
     */
    get placeholderPath() {
        return this.cqPath && this.cqPath + '/' + PLACEHOLDER_ITEM_NAME;
    }
}
AEMContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'aem-container',
                host: {
                    '[class]': 'hostClasses',
                    '[attr.data-cq-data-path]': 'cqPath'
                },
                template: `<ng-container *ngFor="let itemKey of cqItemsOrder">
  <ng-container [aemComponent] [cqItem]='getItem(itemKey)' [cqPath]='getDataPath(itemKey)' [itemName]='itemKey'></ng-container>
  <div *ngIf="isInEditMode" [attr.data-cq-data-path]="placeholderPath" [class]="getPlaceholderClassNames()"></div>
</ng-container>
`
            },] },
];
/** @nocollapse */
AEMContainerComponent.propDecorators = {
    "cqItems": [{ type: Input },],
    "cqItemsOrder": [{ type: Input },],
    "cqPath": [{ type: Input },],
    "modelName": [{ type: Input },],
    "classNames": [{ type: Input },],
};
function AEMContainerComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AEMContainerComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AEMContainerComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    AEMContainerComponent.propDecorators;
    /**
     * Map of model items included in the current container
     * @type {?}
     */
    AEMContainerComponent.prototype.cqItems;
    /**
     * Array of model item keys
     * @type {?}
     */
    AEMContainerComponent.prototype.cqItemsOrder;
    /**
     * Path to the model associated with the current instance of the component
     * @type {?}
     */
    AEMContainerComponent.prototype.cqPath;
    /**
     * Key of the model structure
     * @type {?}
     */
    AEMContainerComponent.prototype.modelName;
    /**
     * Class names of the current component
     * @type {?}
     */
    AEMContainerComponent.prototype.classNames;
}
