// @adobe/cq-angular-editable-components/esm5/lib/layout/aem-model-provider/aem-model-provider.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/*
 * ADOBE CONFIDENTIAL
 *
 * Copyright 2018 Adobe Systems Incorporated
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Adobe Systems Incorporated and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Adobe Systems Incorporated and its
 * suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Adobe Systems Incorporated.
 */
import { Component, Input, NgZone } from '@angular/core';
import { ModelManager } from "@adobe/cq-spa-page-model-manager";
var AEMModelProviderComponent = /** @class */ (function () {
    function AEMModelProviderComponent(ngZone) {
        this.ngZone = ngZone;
    }
    /**
     * Updates the item data
     */
    /**
     * Updates the item data
     * @return {?}
     */
    AEMModelProviderComponent.prototype.updateItem = /**
     * Updates the item data
     * @return {?}
     */
    function () {
        var _this = this;
        ModelManager.getData({ path: this.cqPath }).then(function (model) {
            _this.ngZone.run(function () {
                _this.cqItem = model;
            });
        });
    };
    /**
     * @return {?}
     */
    AEMModelProviderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        ModelManager.addListener(this.cqPath, this.updateItem.bind(this));
    };
    /**
     * @return {?}
     */
    AEMModelProviderComponent.prototype.ngDestroy = /**
     * @return {?}
     */
    function () {
        ModelManager.removeListener(this.cqPath, this.updateItem.bind(this));
    };
    AEMModelProviderComponent.decorators = [
        { type: Component, args: [{
                    selector: '[aemModelProvider]',
                    template: "<ng-container [aemComponent] [cqItem]='cqItem' [cqPath]='cqPath' [itemName]='itemName'></ng-container>"
                },] },
    ];
    /** @nocollapse */
    AEMModelProviderComponent.ctorParameters = function () { return [
        { type: NgZone, },
    ]; };
    AEMModelProviderComponent.propDecorators = {
        "cqPath": [{ type: Input },],
        "cqItem": [{ type: Input },],
        "itemName": [{ type: Input },],
        "aemModelProvider": [{ type: Input },],
    };
    return AEMModelProviderComponent;
}());
export { AEMModelProviderComponent };
function AEMModelProviderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AEMModelProviderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AEMModelProviderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    AEMModelProviderComponent.propDecorators;
    /**
     * Path to the model associated with the current instance of the component
     * @type {?}
     */
    AEMModelProviderComponent.prototype.cqPath;
    /**
     * Model item associated with the current model provider component
     * @type {?}
     */
    AEMModelProviderComponent.prototype.cqItem;
    /**
     * Name of the item associated with the current model provider component
     * @type {?}
     */
    AEMModelProviderComponent.prototype.itemName;
    /** @type {?} */
    AEMModelProviderComponent.prototype.aemModelProvider;
    /** @type {?} */
    AEMModelProviderComponent.prototype.ngZone;
}
